packages/unified-latex-to-pretext/libs/pre-conversion-subs/break-on-boundaries.ts

```diff
@@ -0,0 +1,144 @@
+import { env, arg } from "@unified-latex/unified-latex-builder";
+import * as Ast from "@unified-latex/unified-latex-types";
+import { getNamedArgsContent } from "@unified-latex/unified-latex-util-arguments";
+import {
+    anyEnvironment,
+    anyMacro,
+    match,
+} from "@unified-latex/unified-latex-util-match";
+import { printRaw } from "@unified-latex/unified-latex-util-print-raw";
+import { replaceNode } from "@unified-latex/unified-latex-util-replace";
+import {
+    splitOnMacro,
+    unsplitOnMacro,
+} from "@unified-latex/unified-latex-util-split";
+import { visit } from "@unified-latex/unified-latex-util-visit";
+
+/**
+ * All the divisions, where each item is {division macro, mapped environment}.
+ * Note that this is ordered from the "largest" division to the "smallest" division.
+ */
+const divisions: { division: string; mappedEnviron: string }[] = [
+    { division: "part", mappedEnviron: "_part" },
+    { division: "chapter", mappedEnviron: "_chapter" },
+    { division: "section", mappedEnviron: "_section" },
+    { division: "subsection", mappedEnviron: "_subsection" },
+    { division: "subsubsection", mappedEnviron: "_subsubsection" },
+    { division: "paragraph", mappedEnviron: "_paragraph" },
+    { division: "subparagraph", mappedEnviron: "_subparagraph" },
+];
+
+// check if a macro is a division macro
+const isDivisionMacro = match.createMacroMatcher(
+    divisions.map((x) => x.division)
+);
+
+// check if an environment is a newly created environment
+const isMappedEnviron = match.createEnvironmentMatcher(
+    // doesn't seem to work
+    divisions.map((x) => x.mappedEnviron)
+);
+
+/**
+ * Breaks up division macros into environments. Returns a list of warning messages
+ * for any groups that were removed.
+ */
+export function breakOnBoundaries(ast: Ast.Ast): { messages: string[] } {
+    // messages for any groups removed
+    const messagesLst: { messages: string[] } = { messages: [] };
```

Comment:
Let's go ahead and make this a standard `vfile-message` list of messages, so it will be compatible with other messages we produce. https://www.npmjs.com/package/vfile-message